cub/cub/detail/launcher/cuda_driver.cuh

```diff
@@ -61,54 +61,70 @@ struct CudaDriverLauncher
     }
     else
 #endif
     {
-      return static_cast<cudaError_t>(cuLaunchKernel(
+      return static_cast<::cudaError_t>(::cuLaunchKernel(
         kernel_fn, grid.x, grid.y, grid.z, block.x, block.y, block.z, shared_mem, stream, kernel_args, 0));
     }
   }
 };
 
 struct CudaDriverLauncherFactory
 {
-  CudaDriverLauncher
-  operator()(dim3 grid, dim3 block, _CUDA_VSTD::size_t shared_mem, CUstream stream, bool dependent_launch = false) const
+  CudaDriverLauncher operator()(
+    dim3 grid, dim3 block, _CUDA_VSTD::size_t shared_mem, ::CUstream stream, bool dependent_launch = false) const
   {
     return CudaDriverLauncher{grid, block, shared_mem, stream, dependent_launch};
   }
 
-  cudaError_t PtxVersion(int& version) const
+  ::cudaError_t PtxVersion(int& version) const
   {
     version = cc;
     return cudaSuccess;
   }
 
-  _CCCL_HIDE_FROM_ABI cudaError_t MultiProcessorCount(int& sm_count) const
+  _CCCL_HIDE_FROM_ABI ::cudaError_t MultiProcessorCount(int& sm_count) const
   {
-    return static_cast<cudaError_t>(cuDeviceGetAttribute(&sm_count, CU_DEVICE_ATTRIBUTE_MULTIPROCESSOR_COUNT, device));
+    return static_cast<::cudaError_t>(
+      ::cuDeviceGetAttribute(&sm_count, ::CU_DEVICE_ATTRIBUTE_MULTIPROCESSOR_COUNT, device));
   }
 
-  _CCCL_HIDE_FROM_ABI cudaError_t
-  MaxSmOccupancy(int& sm_occupancy, CUkernel kernel, int block_size, int dynamic_smem_bytes = 0) const
+  _CCCL_HIDE_FROM_ABI ::cudaError_t
+  MaxSmOccupancy(int& sm_occupancy, ::CUkernel kernel, int block_size, int dynamic_smem_bytes = 0) const
   {
     // Older drivers have issues handling CUkernel in the occupancy queries, get the CUfunction instead.
-    CUfunction kernel_fn;
-    auto status = static_cast<cudaError_t>(cuKernelGetFunction(&kernel_fn, kernel));
+    ::CUfunction kernel_fn;
+    auto status = static_cast<::cudaError_t>(::cuKernelGetFunction(&kernel_fn, kernel));
     if (status != cudaSuccess)
     {
       return status;
     }
 
-    return static_cast<cudaError_t>(
-      cuOccupancyMaxActiveBlocksPerMultiprocessor(&sm_occupancy, kernel_fn, block_size, dynamic_smem_bytes));
+    return static_cast<::cudaError_t>(
+      ::cuOccupancyMaxActiveBlocksPerMultiprocessor(&sm_occupancy, kernel_fn, block_size, dynamic_smem_bytes));
   }
 
-  _CCCL_HIDE_FROM_ABI cudaError_t MaxGridDimX(int& max_grid_dim_x) const
+  _CCCL_HIDE_FROM_ABI ::cudaError_t MaxGridDimX(int& max_grid_dim_x) const
   {
-    return static_cast<cudaError_t>(cuDeviceGetAttribute(&max_grid_dim_x, CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_X, device));
+    return static_cast<::cudaError_t>(
+      ::cuDeviceGetAttribute(&max_grid_dim_x, ::CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_X, device));
+  }
+
+  _CCCL_HIDE_FROM_ABI ::cudaError_t
+  MemsetAsync(void* dst, int value, size_t num_elements, size_t /*element_size*/, ::CUstream stream) const
+  {
+    return static_cast<::cudaError_t>(
+      ::cuMemsetD32Async(reinterpret_cast<::CUdeviceptr>(dst), value, num_elements, stream));
+  }
```

Comment:
Important: this API design is very fragile. Consider:
```c++
launcher.MemsetAsync(ptr, 0x12, 1000, 1, stream):
````

Will write 1000 bytes with the pattern `0x121212121212...` when the CUDA runtime launcher is used, and will write 4000 bytes with the pattern `0x000000120000001200000012...` when the driver launcher is used.

I would drop the `element_size` parameter, and change the type of `value` to `unsigned char`. In the driver API you can still call `cuMemsetD32Async` with `value | value << 8 | value << 16 | value << 24`. Furthermore, you should add a check that `num_elements % 4 == 0` and return an error otherwise.

Does this API need to support memsetting buffers with sizes that are not a multiple of 4 bytes? If yes, we have a different problem as well :)

Reply:
I changed the API to accept `unsigned char` as the value instead of `int`. I also got rid of `element_size` and switched to `cuMemsetD8Async`. Now we always assume that the value being set is a byte. This way the two implementations are consistent.